alf/bin/grid_search.py

```diff
@@ -125,21 +174,34 @@ def run(self):
 
         process_pool = multiprocessing.Pool(
             processes=max_worker_num, maxtasksperchild=1)
+        device_queue = self._init_device_queue(max_worker_num)
 
         task_count = 0
         for values in itertools.product(*param_values):
             parameters = dict(zip(param_keys, values))
             root_dir = "%s/%d" % (FLAGS.root_dir, task_count)
             process_pool.apply_async(
                 func=self._worker,
-                args=[root_dir, parameters],
+                args=[root_dir, parameters, device_queue],
                 error_callback=lambda e: logging.error(e))
             task_count += 1
 
         process_pool.close()
         process_pool.join()
 
-    def _worker(self, root_dir, parameters):
+    def _worker(self, root_dir, parameters, device_queue):
+        # sleep for random seconds to avoid crowded launching
+        time.sleep(random.uniform(0, 3))
+
+        device = device_queue.get()
+        if self._conf.use_gpu:
+            os.environ["CUDA_VISIBLE_DEVICES"] = str(device)
```

Comment:
Did you verify that changing CUDA_VISIBLE_DEVICES has effect after the process is started?

Reply:
Yes, as long as there hasn't been no long tf message output (e.g., the long message after you create a tensor for the first time)